Added redis_data module

paginabianca opened the pull request:
Added redis_set module and unit tests for the new module.
Applied suggested changes and removed redis_del in favor of state
option.
Also added redis utility class that handles connection.

##### SUMMARY
- New Module Pull Request

##### COMPONENT NAME
redis_set

##### ADDITIONAL INFORMATION

felixfontein commented:
I would suggest we keep this PR waiting until #3227 is ready for merging. Otherwise we'll have all the discussions on the common code multiple times.

felixfontein commented on plugins/modules/database/misc/redis_set.py:
```diff
@@ -0,0 +1,228 @@
+
+DOCUMENTATION = '''
+---
+module: redis_set
```
Have you thought about renaming this module to `redis_data`, to make clear it belongs to `redis_data_info`? After all it can not just set values (`state=present`), but also delete values (`state=absent`).

felixfontein commented on plugins/modules/database/misc/redis_set.py:
```diff
@@ -0,0 +1,228 @@
+    key:
+        description:
+            - Database key.
+        required: True
```
```suggestion
        required: true
```
YAML booleans should always be lowercase. (Same for all others in this file.)

felixfontein commented on plugins/modules/database/misc/redis_set.py:
```diff
@@ -0,0 +1,228 @@
+    key: foo
+    value: bar
+
+- name: Delete key foo=bar on localhost with no username
```
```suggestion
- name: Delete key foo on localhost with no username
```
It will also be deleted if it has another value than bar, right?

felixfontein commented on plugins/modules/database/misc/redis_set.py:
```diff
@@ -0,0 +1,228 @@
+    login_host: redishost
+    login_password: supersecret
+    login_user: someuser
+    validate_certs: yes
```
```suggestion
    validate_certs: true
```

felixfontein commented on plugins/modules/database/misc/redis_set.py:
```diff
@@ -0,0 +1,228 @@
+  type: str
+  sample: 'old_value_of_key'
+value:
+  description: Value key was set to
```
```suggestion
  description: Value key was set to.
```

felixfontein commented on plugins/modules/database/misc/redis_set.py:
```diff
@@ -0,0 +1,228 @@
+                    key, value)
+            result['msg'] = msg
+            module.fail_json(**result)
+        msg = 'Set key: {0} to {1}'.format(key, value)
```
```suggestion
        msg = 'Set key: {0} to {1}'.format(key, value)
```
I would *really* avoid putting the value into the message as well. It's already returned as a separate result.

paginabianca commented:
Hi :wave: 
I applied the suggested changes and removed the value from the return messages. I also fixed some inconsistencies with the description of the returned values. For example deleting a value will not return `value` because on success the value would always be None anyway.

felixfontein commented on plugins/modules/database/misc/redis_data.py:
```diff
@@ -0,0 +1,226 @@
+        msg = 'Key {0} already has desired value'.format(key)
+        result['msg'] = msg
+        result['value'] = value
+        module.exit_json(**result)
```
What about all the other options? For idempotency, they also need to be compared (if provided) and updated.

paginabianca replied:
I am not sure if it is possible to have full idempotency for the other values. The only option that would make sense is expiration time, but Redis only provides the UNIT time stamp when the key will expire so there is no way of knowing when the key/the expiry of the key was first set.
This means when the `expiration` option is set, the module is never idempotent.

felixfontein replied:
In that case, the documentation should be very clear what's used for idempotency and what is not.

felixfontein commented on plugins/modules/database/misc/redis_data.py:
```diff
@@ -0,0 +1,226 @@
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=False,
```
It would be nice to also implement check mode.

felixfontein commented on plugins/modules/database/misc/redis_data.py:
```diff
@@ -0,0 +1,249 @@
+        module.fail_json(**result)
+
+    result['old_value'] = old_value
+    if ((old_value == value) and ((keepttl is not None and keepttl) or px is None)):
```
Shouldn't it be `and px is None`, since if it is not `None`, according to the docs the status should *always* be changed?
```suggestion
    if (old_value == value) and (keepttl is not None and keepttl) and px is None:
```
Also I'm not sure what to think of `keepttl is not None and keepttl`. That's basically equivalent to `keepttl` or `keepttl is True`. But if `keepttl` is not specified, shouldn't it also be ignored? I.e.
```suggestion
    if old_value == value and keepttl is not False and px is None:
```

paginabianca replied:
You are right. I made a mess on this one. The status should always change when `px` is set.

felixfontein reviewed (approved):
LGTM

felixfontein commented:
@paginabianca thanks for contributing this module!

paginabianca commented:
Awesome! Thanks for the support and feedback :pray:
